subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/DefaultRootComponentMetadataBuilderTest.groovy

```diff
@@ -0,0 +1,101 @@
+import org.gradle.api.internal.artifacts.Module
+import org.gradle.api.internal.artifacts.component.ComponentIdentifierFactory
+import org.gradle.api.internal.artifacts.configurations.ConfigurationsProvider
+import org.gradle.api.internal.artifacts.configurations.DependencyMetaDataProvider
+import org.gradle.api.internal.artifacts.configurations.MutationValidator
+import org.gradle.api.internal.artifacts.dsl.dependencies.ProjectFinder
+import org.gradle.internal.component.external.model.DefaultModuleComponentIdentifier
+import spock.lang.Specification
+import spock.lang.Unroll
+
+class DefaultRootComponentMetadataBuilderTest extends Specification {
+
+    DependencyMetaDataProvider metaDataProvider = Mock() {
+        getModule() >> Mock(Module)
+    }
+    ComponentIdentifierFactory componentIdentifierFactory = Mock()
+    ImmutableModuleIdentifierFactory moduleIdentifierFactory = Mock()
+    ProjectFinder projectFinder = Mock()
+    ConfigurationComponentMetaDataBuilder configurationComponentMetaDataBuilder = Mock()
+    ConfigurationsProvider configurationsProvider = Mock()
+
+    def builder = new DefaultRootComponentMetadataBuilder(
+        metaDataProvider,
+        componentIdentifierFactory,
+        moduleIdentifierFactory,
+        projectFinder,
+        configurationComponentMetaDataBuilder,
+        configurationsProvider)
+
+    def "caches root component metadata"() {
+        componentIdentifierFactory.createComponentIdentifier(_) >> {
+            new DefaultModuleComponentIdentifier('foo', 'bar', '1.0')
+        }
+        def root = builder.toRootComponentMetaData()
+
+        when:
+        def otherRoot = builder.toRootComponentMetaData()
+
+        then:
+        otherRoot.is(root)
+    }
+
+    def "doesn't cache root component metadata when module identifier changes"() {
+        1 * componentIdentifierFactory.createComponentIdentifier(_) >> {
+            new DefaultModuleComponentIdentifier('foo', 'bar', '1.0')
+        }
+        def root = builder.toRootComponentMetaData()
+
+        when:
+        componentIdentifierFactory.createComponentIdentifier(_) >> {
+            new DefaultModuleComponentIdentifier('foo', 'baz', '1.0')
+        }
+
+        def otherRoot = builder.toRootComponentMetaData()
+
+        then:
+        !otherRoot.is(root)
+    }
+
+    @Unroll
+    def "caching of component metadata when #mutationType change"() {
+        componentIdentifierFactory.createComponentIdentifier(_) >> {
+            new DefaultModuleComponentIdentifier('foo', 'bar', '1.0')
+        }
+        def root = builder.toRootComponentMetaData()
+
+        when:
+        builder.validator.validateMutation(mutationType)
+        def otherRoot = builder.toRootComponentMetaData()
+
+        then:
+        otherRoot.is(root) == cached
+
+        where:
+        mutationType                                | cached
+        MutationValidator.MutationType.DEPENDENCIES | false
```

Comment:
Attributes are locked during the root component metadata creation. It's only unfortunate that we allow mutation of configurations _after_ root component metadata has been created. It should be an error, but for backwards compatibility, we cannot enforce it like we did for attributes.